src/IR/DConToFunc.hs

```diff
@@ -114,9 +114,89 @@ createArityEnv defs = M.fromList $ concatMap arities defs
 dConToFunc :: I.Program Poly.Type -> Compiler.Pass (I.Program Poly.Type)
 dConToFunc p@I.Program { I.programDefs = defs, I.typeDefs = tDefs } =
   runArityFn (createArityEnv tDefs) $ do
-    defs' <- everywhereM (mkM dataToApp) defs
-    return p { I.programDefs = defs' ++ concat (createFuncs <$> tDefs) }
+    defs'' <- defs' -- user defined functions
+    return p { I.programDefs = tDefs' ++ defs'' } -- constructor funcs ++ user funcs
  where
+  tDefs' = concat (createFuncs <$> tDefs)
+  -- ^ top-level constructor functions
+  defs' =
+    filter ((`notElem` (fst <$> tDefs')) . fst)
+      <$> everywhereM (mkM dataToApp) defs
+  -- ^ We need to filter defs to account for name collisions
+  -- ^ between user defined funcs and newly created and inserted constructor funcs
+  {- Example of name collision case and correction
+  Given the SSLANG program
+  @
+  type Color
+    White
+    Black
+    RGB Int Int Int
+
+  main ( cout : & Int ) -> () = ()
+  @
+  dConToFunc will produce
+  @
+  type Color
+    White
+    Black
+    RGB Int Int Int
+
+  __RGB (__arg0 : Int) (__arg1 : Int) (__arg2 : Int) -> Color = 
+  RGB __arg0 __arg1 __arg2
+    
+  main ( cout : & Int ) -> () = ()
+  @
+  This is okay.
+  Now, given the SSLANG program
+  @
+  type Color
+    White
+    Black
+    RGB Int Int Int
+
+  __RGB (r : Int) (g : Int) (b : Int) -> Color = 
```

Comment:
I would argue that the user should not be able to write this sslang program

Reply:
**Would you like me to address this issue before merging this PR into main?** I could add a check in `dcontofunc` that throws an error on finding name collisions with user defined functions, but this will break our pretty printer tests. I talked a bit more about this here: (https://github.com/ssm-lang/sslang/pull/92#discussion_r907892990)